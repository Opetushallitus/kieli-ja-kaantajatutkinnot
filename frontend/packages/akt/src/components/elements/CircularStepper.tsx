import { CircularProgress, CircularProgressProps } from '@mui/material';

import { H3 } from 'components/elements/Text';
import { useCommonTranslation } from 'configs/i18n';
import { Color } from 'enums/app';

interface CircularStepperProps extends CircularProgressProps {
  value: number;
  phaseText: string;
  phaseDescription: string;
}

export const CircularStepper = ({
  phaseText,
  phaseDescription,
  ...rest
}: CircularStepperProps) => {
  const translateCommon = useCommonTranslation();
  const ariaLabel = `${translateCommon(
    'phase'
  )} ${phaseText}: ${phaseDescription}`;

  return (
    <div className="circular-stepper">
      <CircularProgress
        aria-label={ariaLabel}
        className="circular-stepper__progress"
        color={Color.Secondary}
        variant="determinate"
        {...rest}
      />
      <div className="circular-stepper__heading">
        <H3>{phaseText}</H3>
      </div>
    </div>
  );
};
