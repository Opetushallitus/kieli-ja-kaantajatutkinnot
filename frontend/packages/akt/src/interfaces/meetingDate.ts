import { Dayjs } from 'dayjs';
import { Action } from 'redux';

import { APIResponseStatus } from 'enums/api';
import { MeetingStatus } from 'enums/meetingDate';
import { WithId, WithVersion } from 'interfaces/with';
import {
  MEETING_DATE_ADD,
  MEETING_DATE_REMOVE,
} from 'redux/actionTypes/meetingDate';

export interface MeetingDates {
  meetingDates: Array<MeetingDate>;
}

export interface MeetingDateFilter {
  meetingStatus: MeetingStatus;
}
export interface MeetingDatesState {
  status: APIResponseStatus;
  filters: MeetingDateFilter;
  meetingDates: MeetingDate[];
}

export interface AddMeetingDateState {
  status: APIResponseStatus;
  date: Dayjs;
}

export interface MeetingDateState {
  meetingDates: MeetingDatesState;
  addMeetingDate: AddMeetingDateState;
  removeMeetingDate: RemoveMeetingDateState;
}

export interface MeetingDateAction
  extends Action<string>,
    Partial<MeetingDates> {
  filters?: MeetingDateFilter;
  meetingDateId?: number;
  status?: APIResponseStatus;
  date?: Dayjs;
}

export interface MeetingDate extends Omit<MeetingDateResponse, 'date'> {
  date: Dayjs;
}
export interface MeetingDateResponse extends WithId, WithVersion {
  date: string;
}
export interface RemoveMeetingDateState {
  status: APIResponseStatus;
  meetingDateId: number | undefined;
}

export interface RemoveMeetingDateAction extends Action<string> {
  meetingDateId: number;
}

export interface AddMeetingDateAction extends Action<string> {
  date: Dayjs;
}

export type AddMeetingDateActionType = {
  type: typeof MEETING_DATE_ADD;
  date: Dayjs;
};

export type RemoveMeetingDateActionType = {
  type: typeof MEETING_DATE_REMOVE;
  meetingDateId: number;
};
