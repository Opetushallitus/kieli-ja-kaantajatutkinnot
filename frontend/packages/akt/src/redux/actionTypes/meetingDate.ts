export const MEETING_DATE_LOAD = 'MEETING_DATE/LOAD';
export const MEETING_DATE_LOADING = 'MEETING_DATE/LOADING';
export const MEETING_DATE_ERROR = 'MEETING_DATE/ERROR';
export const MEETING_DATE_RECEIVED = 'MEETING_DATE/RECEIVED';

export const MEETING_DATE_ADD_FILTER = 'MEETING_DATE/ADD_FILTER';

export const MEETING_DATE_ADD = 'MEETING_DATE/ADD';
export const MEETING_DATE_ADD_SUCCESS = 'MEETING_DATE/ADD_SUCCESS';
export const MEETING_DATE_ADD_ERROR = 'MEETING_DATE/ADD_ERROR';

export const MEETING_DATE_REMOVE = 'MEETING_DATE/REMOVE';
export const MEETING_DATE_REMOVE_SUCCESS = 'MEETING_DATE/REMOVE_SUCCESS';
export const MEETING_DATE_REMOVE_ERROR = 'MEETING_DATE/REMOVE_ERROR';

export const MEETING_DATE_REMOVE_CONFIRM = 'MEETING_DATE/REMOVE';
export const MEETING_DATE_REMOVE_CANCEL = 'MEETING_DATE/CANCEL';
