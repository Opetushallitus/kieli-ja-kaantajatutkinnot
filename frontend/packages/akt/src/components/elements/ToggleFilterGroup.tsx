import { CustomButton } from 'components/elements/CustomButton';
import { Color, Variant } from 'enums/app';

interface ToggleFilter<T> {
  status: T;
  label: string;
  count?: number;
  testId?: string;
}
interface ToggleFilterGroupProps<T> {
  filters: Array<ToggleFilter<T>>;
  activeStatus: T;
  onButtonClick(T: T): void;
}

export function ToggleFilterGroup<T>({
  filters,
  activeStatus,
  onButtonClick,
}: ToggleFilterGroupProps<T>) {
  return (
    <div className="columns">
      {filters.map(({ count, status, testId, label }, i) => (
        <CustomButton
          className="border-radius-unset"
          key={i}
          data-testid={testId}
          color={Color.Secondary}
          variant={
            activeStatus === status ? Variant.Contained : Variant.Outlined
          }
          onClick={() => onButtonClick(status)}
        >
          <div className="columns gapped">
            <div className="grow">{label}</div>
            {count && <div>{`(${count})`}</div>}
          </div>
        </CustomButton>
      ))}
    </div>
  );
}
