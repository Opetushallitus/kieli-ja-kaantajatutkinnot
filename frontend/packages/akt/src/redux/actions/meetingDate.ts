import { Dayjs } from 'dayjs';

import { MeetingDateFilter } from 'interfaces/meetingDate';
import {
  MEETING_DATE_ADD,
  MEETING_DATE_ADD_FILTER,
  MEETING_DATE_LOAD,
  MEETING_DATE_REMOVE,
} from 'redux/actionTypes/meetingDate';

export const removeMeetingDate = (meetingDateId: number) => ({
  type: MEETING_DATE_REMOVE,
  meetingDateId,
});

export const loadMeetingDates = {
  type: MEETING_DATE_LOAD,
};

export const setMeetingDateFilters = (filters: MeetingDateFilter) => ({
  type: MEETING_DATE_ADD_FILTER,
  filters,
});

export const addMeetingDate = (date: Dayjs) => ({
  type: MEETING_DATE_ADD,
  date,
});
