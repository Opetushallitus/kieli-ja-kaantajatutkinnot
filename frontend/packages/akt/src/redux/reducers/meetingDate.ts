import { Dayjs } from 'dayjs';
import { Reducer } from 'redux';

import { APIResponseStatus } from 'enums/api';
import { MeetingStatus } from 'enums/meetingDate';
import {
  MeetingDate,
  MeetingDateAction,
  MeetingDateFilter,
  MeetingDateState,
} from 'interfaces/meetingDate';
import {
  MEETING_DATE_ADD,
  MEETING_DATE_ADD_ERROR,
  MEETING_DATE_ADD_FILTER,
  MEETING_DATE_ADD_SUCCESS,
  MEETING_DATE_ERROR,
  MEETING_DATE_LOADING,
  MEETING_DATE_RECEIVED,
  MEETING_DATE_REMOVE,
  MEETING_DATE_REMOVE_ERROR,
  MEETING_DATE_REMOVE_SUCCESS,
} from 'redux/actionTypes/meetingDate';
import { DateUtils } from 'utils/date';

const dayjs = DateUtils.dayjs();

const defaultState = {
  meetingDates: {
    status: APIResponseStatus.NotStarted,
    meetingDates: [],
    filters: {
      meetingStatus: MeetingStatus.Upcoming,
    },
  },
  addMeetingDate: {
    status: APIResponseStatus.NotStarted,
    date: dayjs(),
  },
  removeMeetingDate: {
    status: APIResponseStatus.NotStarted,
    meetingDateId: undefined,
  },
};

export const meetingDateReducer: Reducer<
  MeetingDateState,
  MeetingDateAction
> = (state = defaultState, action) => {
  const filters = action.filters as MeetingDateFilter;
  switch (action.type) {
    // Fetching meetings
    case MEETING_DATE_LOADING:
      return {
        ...state,
        meetingDates: {
          ...state.meetingDates,
          status: APIResponseStatus.InProgress,
        },
      };
    case MEETING_DATE_RECEIVED:
      const meetingDates = action.meetingDates as Array<MeetingDate>;

      return {
        ...state,
        meetingDates: {
          ...state.meetingDates,
          meetingDates,
          status: APIResponseStatus.Success,
        },
      };
    case MEETING_DATE_ERROR:
      return {
        ...state,
        meetingDates: {
          ...state.meetingDates,
          status: APIResponseStatus.Error,
        },
      };

    case MEETING_DATE_ADD_FILTER:
      return {
        ...state,
        meetingDates: {
          ...state.meetingDates,
          filters: { ...state.meetingDates.filters, ...filters },
        },
      };

    // Removing meetings
    case MEETING_DATE_REMOVE:
      const { meetingDateId } = action;

      return {
        ...state,
        removeMeetingDate: {
          ...state.removeMeetingDate,
          meetingDateId,
          status: APIResponseStatus.InProgress,
        },
      };
    case MEETING_DATE_REMOVE_SUCCESS:
      return {
        ...state,
        removeMeetingDate: {
          ...state.removeMeetingDate,
          status: APIResponseStatus.Success,
        },
      };
    case MEETING_DATE_REMOVE_ERROR:
      return {
        ...state,
        removeMeetingDate: {
          ...state.removeMeetingDate,
          status: APIResponseStatus.Error,
        },
      };

    // Adding meetings
    case MEETING_DATE_ADD:
      const date = action.date as Dayjs;

      return {
        ...state,
        addMeetingDate: {
          ...state.addMeetingDate,
          date,
          status: APIResponseStatus.InProgress,
        },
      };
    case MEETING_DATE_ADD_SUCCESS:
      return {
        ...state,
        addMeetingDate: {
          ...state.addMeetingDate,
          status: APIResponseStatus.Success,
        },
      };
    case MEETING_DATE_ADD_ERROR:
      return {
        ...state,
        addMeetingDate: {
          ...state.addMeetingDate,
          status: APIResponseStatus.Error,
        },
      };

    default:
      return state;
  }
};
