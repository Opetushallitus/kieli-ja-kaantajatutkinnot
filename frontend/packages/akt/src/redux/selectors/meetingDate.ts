import { Dayjs } from 'dayjs';
import { createSelector } from 'reselect';

import { RootState } from 'configs/redux';
import { MeetingStatus } from 'enums/meetingDate';
import { DateUtils } from 'utils/date';

export const meetingDatesSelector = (state: RootState) => state.meetingDate;

export const selectMeetingDatesByMeetingStatus = createSelector(
  (state: RootState) => state.meetingDate.meetingDates,
  (meetingDates) => {
    // TODO Note that this has an *implicit* dependency on the current system time,
    // which we currently fail to take into account properly - the selectors should
    // somehow make the dependency on time explicit!
    const dayjs = DateUtils.dayjs();
    const now = dayjs();
    const upcoming = meetingDates.meetingDates
      .filter(({ date }) =>
        filterMeetingDateByStatus(date, MeetingStatus.Upcoming, now)
      )
      .sort((a, b) => a.date.valueOf() - b.date.valueOf());

    const passed = meetingDates.meetingDates
      .filter(({ date }) =>
        filterMeetingDateByStatus(date, MeetingStatus.Passed, now)
      )
      .sort((a, b) => b.date.valueOf() - a.date.valueOf());

    return {
      upcoming,
      passed,
    };
  }
);

const filterMeetingDateByStatus = (
  date: Dayjs,
  status: MeetingStatus,
  currentDate: Dayjs
) => {
  const isBefore = DateUtils.isDatePartBefore(date, currentDate);

  return status === MeetingStatus.Upcoming ? !isBefore : isBefore;
};
