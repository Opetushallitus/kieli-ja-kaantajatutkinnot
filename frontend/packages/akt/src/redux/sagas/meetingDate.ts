import { AxiosError, AxiosResponse } from 'axios';
import { call, put, takeLatest } from 'redux-saga/effects';

import axiosInstance from 'configs/axios';
import { APIEndpoints } from 'enums/api';
import {
  AddMeetingDateActionType,
  MeetingDateResponse,
  MeetingDates,
  RemoveMeetingDateActionType,
} from 'interfaces/meetingDate';
import {
  MEETING_DATE_ADD,
  MEETING_DATE_ADD_ERROR,
  MEETING_DATE_ADD_SUCCESS,
  MEETING_DATE_ERROR,
  MEETING_DATE_LOAD,
  MEETING_DATE_LOADING,
  MEETING_DATE_RECEIVED,
  MEETING_DATE_REMOVE,
  MEETING_DATE_REMOVE_ERROR,
  MEETING_DATE_REMOVE_SUCCESS,
} from 'redux/actionTypes/meetingDate';
import { NOTIFIER_TOAST_ADD } from 'redux/actionTypes/notifier';
import { DateUtils } from 'utils/date';
import { NotifierUtils } from 'utils/notifier';
import { SerializationUtils } from 'utils/serialization';

export function* removeMeetingDate(action: RemoveMeetingDateActionType) {
  try {
    yield call(
      axiosInstance.delete,
      `${APIEndpoints.MeetingDate}/${action.meetingDateId}`
    );
    yield put({ type: MEETING_DATE_REMOVE_SUCCESS });
    yield put({ type: MEETING_DATE_LOAD });
  } catch (error) {
    yield put({ type: MEETING_DATE_REMOVE_ERROR });
    yield put({
      type: NOTIFIER_TOAST_ADD,
      notifier: NotifierUtils.createAxiosErrorNotifierToast(
        error as AxiosError
      ),
    });
  }
}

export function* addMeetingDate(action: AddMeetingDateActionType) {
  try {
    yield call(axiosInstance.post, APIEndpoints.MeetingDate, {
      date: DateUtils.serializeDate(action.date),
    });
    yield put({ type: MEETING_DATE_ADD_SUCCESS });
    yield put({ type: MEETING_DATE_LOAD });
  } catch (error) {
    yield put({ type: MEETING_DATE_ADD_ERROR });
    yield put({
      type: NOTIFIER_TOAST_ADD,
      notifier: NotifierUtils.createAxiosErrorNotifierToast(
        error as AxiosError
      ),
    });
  }
}

function* fetchMeetingDates() {
  try {
    yield put({ type: MEETING_DATE_LOADING });
    const apiResponse: AxiosResponse<Array<MeetingDateResponse>> = yield call(
      axiosInstance.get,
      APIEndpoints.MeetingDate
    );

    const deserializedResponse = SerializationUtils.deserializeMeetingDates(
      apiResponse.data
    );
    yield call(storeApiResults, deserializedResponse);
  } catch (error) {
    yield put({ type: MEETING_DATE_ERROR, error });
  }
}

export function* storeApiResults(response: MeetingDates) {
  const { meetingDates } = response;

  yield put({
    type: MEETING_DATE_RECEIVED,
    meetingDates,
  });
}

export function* watchMeetingDates() {
  yield takeLatest(MEETING_DATE_LOAD, fetchMeetingDates);
  yield takeLatest(MEETING_DATE_ADD, addMeetingDate);
  yield takeLatest(MEETING_DATE_REMOVE, removeMeetingDate);
}
