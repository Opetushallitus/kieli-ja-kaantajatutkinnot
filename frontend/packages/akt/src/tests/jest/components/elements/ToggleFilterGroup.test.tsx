import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { ToggleFilterGroup } from 'components/elements/ToggleFilterGroup';
import { MeetingStatus } from 'enums/meetingDate';
import { meetingDateToggleFilters } from 'tests/jest/__fixtures__/toggleFilters';

describe('ToggleFilter', () => {
  it('should render ToggleFilter correctly', () => {
    const tree = renderer
      .create(
        <ToggleFilterGroup
          filters={meetingDateToggleFilters}
          activeStatus={MeetingStatus.Upcoming}
          onButtonClick={jest.fn()}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should call "onClick" prop on button click', () => {
    const onClick = jest.fn();
    const { getByText } = render(
      <ToggleFilterGroup
        filters={meetingDateToggleFilters}
        activeStatus={MeetingStatus.Upcoming}
        onButtonClick={onClick}
      />
    );

    fireEvent.click(getByText(/label2/i));

    expect(onClick).toHaveBeenCalled();
  });
});
