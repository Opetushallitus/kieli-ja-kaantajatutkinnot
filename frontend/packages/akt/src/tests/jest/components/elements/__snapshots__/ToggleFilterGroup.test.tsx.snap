// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleFilter should render ToggleFilter correctly 1`] = `
<div
  className="columns"
>
  <button
    className="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root border-radius-unset css-zcbmsk-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <div
      className="columns gapped"
    >
      <div
        className="grow"
      >
        label
      </div>
      <div>
        (10)
      </div>
    </div>
  </button>
  <button
    className="MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root border-radius-unset css-wnpvqa-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <div
      className="columns gapped"
    >
      <div
        className="grow"
      >
        label2
      </div>
      <div>
        (20)
      </div>
    </div>
  </button>
</div>
`;
